doc/langdef.md

```diff
@@ -646,26 +646,42 @@ macros are:
+The semantics depends on the type of the result of evaluating expression `e`:
 
-1.  If `e` is a message and `f` is not declared in this message, the runtime
+1.  If `e` evaluates to a message and `f` is not declared in this message, the runtime
     error `no_such_field` is raised.
-2.  If `e` is a message and `f` is declared, but the field is not set, the
+2.  If `e` evaluates to a message and `f` is declared, but the field is not set, the
     default value of the field's type will be produced. Note that this is `null`
     for messages or the according primitive default value as determined by
     proto2 or proto3 semantics.
-3.  If `e` is a map and `f` is not present in the map, a runtime error will be
-    produced. (Note the runtime error is not a well-known one like
-    `no_such_field` but implementation dependent.) It holds that `e.f ==
-    e['f']`.
-
-To test for the presence of a field, the macro `has(e.f)` can be used.
-`has(e.f)` behaves similar as `e.f`, except as where the former would produce
-`null` or an error different than `no_such_field`, it will return false, and
-true otherwise. This means means that `has(e.f)` applied to a message which does
-not declare field `f` produces a `no_such_field` error, where it produces false
-if `f` is declared but not set (or, in proto3, has its default value). Moreover,
-`has(e.f)` where `e` is a map returns false if `f` is not defined in the map.
+3.  If `e` evaluates to a map, then `e.f` is equivalent to `e['f']` (where `f`
+    is still being used as a meta-variable, e.g. the expression `x.foo` is
+    equivalent to the expression `x['foo']` when `x` evaluates to a map).
+4.  In all other cases, `e.f` evaluates to an error.
+
+To test for the presence of a field, the boolean-valued macro `has(e.f)` can be
+used.
+
+1.  If `e` evaluates to a map, then `has(e.f)` indicates whether the string `f`
+    is a key in the map (note that `f` must syntactically be an identifier).
+2.  If `e` evaluates to a message and `f` is not a declared field for the
+    message, `has(e.f)` is `false`.
+3.  If `e` evaluates to a protocol buffers version 2 message and `f` is a
+    defined field:
+    -   If `f` is a repeated field or map field, `has(e.f)` indicates whether
+        the field is non-empty.
+    -   If `f` is a singular or  oneof field, `has(e.f)` indicates
+        whether the field is set.
+4.  If `e` evaluates to a protocol buffers version 3 message and `f` is a
+    defined field:
+    -   If `f` is a repeated field or map field, `has(e.f)` indicates whether
+        the field is non-empty.
+    -   If `f` is a oneof or singular message field, `has(e.f)` indicates
+        whether the field is set.
+    -   If `f` is some other singular field, `has(e.f)` indicates whether the
+        field's value is its default value (zero for numeric fields, false for
+        booleans, empty for strings and bytes).
+5.  In all other cases, `has(e.f)` evaluates to an error.
```

Comment:
Perhaps it's just worth saying that has(e.f) is not support outside these cases? Typically, you'll get a check error rather than an evaluation error. It's a subtle distinction, but I want to communicate that the error wouldn't necessarily be at eval time.

Reply:
Ah, this reveals a bug in my description: `has(e.undefined_field)` should evaluate to `no_such_field` error instead of `false`. We describe that clearly in an example in the original, but the example contradicts the sentence preceding it.

I've fixed the description of the evaluation logic. Rather than flag all possible concrete errors that the checker might find, I've updated the Gradual Type Checking section to
make it clear that the checker attempts to foresee any runtime errors, not just `no_such_field` and `no_matching_overload`, which implies that `has(e.f)` might fail in the checker when `e` doesn't evaluate to a map or message.

PTAL.

Reply:
LGTM